clustermesh-apiserver/vmmanager.go

```diff
@@ -29,74 +30,107 @@ import (
 	ciliumv2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
 	k8sClient "github.com/cilium/cilium/pkg/k8s/client"
 	clientset "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned"
-	"github.com/cilium/cilium/pkg/k8s/informer"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	"github.com/cilium/cilium/pkg/k8s/synced"
 	"github.com/cilium/cilium/pkg/kvstore"
 	"github.com/cilium/cilium/pkg/kvstore/store"
 	"github.com/cilium/cilium/pkg/labels"
 	nodeStore "github.com/cilium/cilium/pkg/node/store"
 	nodeTypes "github.com/cilium/cilium/pkg/node/types"
 	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/promise"
 )
 
-type VMManager struct {
-	clusterInfo cmtypes.ClusterInfo
+var externalWorkloadsCell = cell.Module(
+	"external-workloads",
+	"External workloads",
 
-	ciliumClient      clientset.Interface
-	identityAllocator *identityCache.CachingIdentityAllocator
+	cell.Config(
+		// The default value is set to true to match the existing behavior in case
+		// the flag is not configured (for instance by the legacy cilium CLI).
+		ExternalWorkloadsConfig{EnableExternalWorkloads: true},
+	),
+
+	cell.Provide(externalWorkloadsProvider),
+	cell.Invoke(func(*VMManager) {}),
+)
 
-	ciliumExternalWorkloadStore    cache.Store
-	ciliumExternalWorkloadInformer cache.Controller
+type ExternalWorkloadsConfig struct {
+	EnableExternalWorkloads bool
+}
 
-	backend kvstore.BackendOperations
+func (def ExternalWorkloadsConfig) Flags(flags *pflag.FlagSet) {
+	flags.Bool("enable-external-workloads", def.EnableExternalWorkloads, "Enable support for external workloads")
 }
 
-func NewVMManager(cinfo cmtypes.ClusterInfo, clientset k8sClient.Clientset, backend kvstore.BackendOperations) *VMManager {
-	m := &VMManager{
-		clusterInfo:  cinfo,
-		ciliumClient: clientset,
-		backend:      backend,
+func externalWorkloadsProvider(
+	lc hive.Lifecycle,
+
+	cfg ExternalWorkloadsConfig,
+	clusterInfo cmtypes.ClusterInfo,
+
+	clientset k8sClient.Clientset,
+	ciliumExternalWorkloads resource.Resource[*ciliumv2.CiliumExternalWorkload],
+	backendPromise promise.Promise[kvstore.BackendOperations],
+) *VMManager {
+	if !cfg.EnableExternalWorkloads {
+		return nil
 	}
-	m.identityAllocator = identityCache.NewCachingIdentityAllocator(m)
 
-	if option.Config.EnableWellKnownIdentities {
-		identity.InitWellKnownIdentities(option.Config, cinfo)
+	// External workloads require CRD allocation mode
+	option.Config.IdentityAllocationMode = option.IdentityAllocationModeCRD
```

Comment:
Does this draw from the regular cilium-config configmap, if so should we terminate if we're using an external kvstore in case Cilium is using that (instead of silently reconfiguring)?

Reply:
The clustermesh-apiserver pod doesn't mount the cilium-config configmap, which means that the IdentityAllocationMode was never set to a value different from the default one also in the current version. From this point of view, no regression is introduced by this PR, which just removes that (unused) parameter and hard-codes the default value for the sake of simplicity.

More in general, the entire clustermesh-apiserver (not only the external workloads support) is not compatible with Cilium running in kvstore mode, and I agree that we should error out in this case to make the incompatibility more explicit (e.g., introducing a validation at the helm level). I'd postpone that as a follow-up item though, being a separate QoL enhancement unrelated from this PR.

Reply:
Resolving the conversation so that this PR can be merged.